File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/util/Request.java
package com.emerchantpay.gateway.genesisandroid.api.util;


import com.emerchantpay.gateway.genesisandroid.api.interfaces.AddressAttributes;
import com.emerchantpay.gateway.genesisandroid.api.interfaces.BaseAttributes;

import java.io.Serializable;

/**
 * Abstract class for fluent interface request builders.
 */
public abstract class Request implements BaseAttributes, AddressAttributes, Serializable {

	public String toXML() {
		throw new UnsupportedOperationException();
	}

	public String toQueryString(String parent) {
		throw new UnsupportedOperationException();
	}

	public String toQueryString() {
		throw new UnsupportedOperationException();
	}

	public String getKind() {
		return null;
	}

	protected String buildXMLElement(Object element) {
		return RequestBuilder.buildXMLElement(element);
	}

	protected String buildXMLElement(String name, Object element) {
		return RequestBuilder.buildXMLElement(name, element);
	}

	public Request getRequest() {
		return this;
	}

	public String getTransactionType() {
		return null;
	}
}

File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/util/Request.kt
package com.emerchantpay.gateway.genesisandroid.api.util


import com.emerchantpay.gateway.genesisandroid.api.interfaces.AddressAttributes
import com.emerchantpay.gateway.genesisandroid.api.interfaces.BaseAttributes

import java.io.Serializable

/**
 * Abstract class for fluent interface request builders.
 */
abstract class Request : BaseAttributes, AddressAttributes, Serializable {

    val kind: String?
        get() = null

    val request: Request
        get() = this

    open fun getTransactionType(): String? {
        return null
    }

    open fun toXML(): String? {
        throw UnsupportedOperationException()
    }

    open fun toQueryString(parent: String): String {
        throw UnsupportedOperationException()
    }

    open fun toQueryString(): String? {
        throw UnsupportedOperationException()
    }

    protected fun buildXMLElement(element: Any): String? {
        return RequestBuilder.buildXMLElement(element)
    }

    protected fun buildXMLElement(name: String, element: Any): String? {
        return RequestBuilder.buildXMLElement(name, element)
    }
}
